nano/node/node.cpp

```diff
@@ -596,17 +617,152 @@ bool nano::network::reachout (nano::endpoint const & endpoint_a, bool allow_loca
 	if (!error)
 	{
 		error |= udp_channels.reachout (endpoint_a);
+		error |= tcp_channels.reachout (endpoint_a);
 	}
 	return error;
 }
 
+std::deque<std::shared_ptr<nano::transport::channel>> nano::network::list (size_t count_a)
+{
+	std::deque<std::shared_ptr<nano::transport::channel>> result;
+	tcp_channels.list (result);
+	udp_channels.list (result);
+	random_pool::shuffle (result.begin (), result.end ());
+	if (result.size () > count_a)
+	{
+		result.resize (count_a, nullptr);
+	}
+	return result;
+}
+
+// Simulating with sqrt_broadcast_simulate shows we only need to broadcast to sqrt(total_peers) random peers in order to successfully publish to everyone with high probability
+std::deque<std::shared_ptr<nano::transport::channel>> nano::network::list_fanout ()
+{
+	auto result (list (size_sqrt ()));
+	return result;
+}
+
+std::unordered_set<std::shared_ptr<nano::transport::channel>> nano::network::random_set (size_t count_a) const
+{
+	std::unordered_set<std::shared_ptr<nano::transport::channel>> result (tcp_channels.random_set (count_a));
+	std::unordered_set<std::shared_ptr<nano::transport::channel>> udp_random (udp_channels.random_set (count_a));
+	for (auto i (udp_random.begin ()), n (udp_random.end ()); i != n && result.size () < count_a * 1.5; ++i)
+	{
+		result.insert (*i);
+	}
+	while (result.size () > count_a)
+	{
+		result.erase (result.begin ());
```

Comment:
Can this use the ranged iterator version of `erase`?

Reply:
Then it will be something like this I guess
```
if (result.size () > count_a)
{
	auto iterator (result.begin ());
	for (auto count (0); iterator != result.end () && count != count_a; ++iterator, ++count)
	{
	}
	result.erase (iterator, result.end ());
	assert (result.size () == count_a);
}
```

Reply:
does this not work?
```
if (result.size () > count_a)
{
result.erase (result.begin (), result.begin () + result.size () - count_a);
}
```
There can be a performance gain removing in bulk for some data structures, `unordered_set` probably not, so don't worry about it too much.

Reply:
Not working this way for unordered_set